### PR TITLE
change half gemm to use compute_32f as default

### PR DESCRIPTION
**Pull request by tianleiwu**
**Description**:  Current fp16 gemm in cuda is aggregated in fp16. It has some problems:
(1) That will introduce loss in precision.
(2) Some older GPU like M60 does not support that.

There are following changes: 
* The default compute type is changed to aggregate in fp32. In this way, it is consistent with training code. 
* Add an environment variable ORT_CUDA_GEMM_OPTIONS for testing purpose.
* Update attention cuda operator to apply the above changes.
* Add unit tests.

Benchmark results (with CUDA 11.0) shows that the latency are very close in most cases, and sometime aggregating in fp16 could be faster:
* HG: cublasHgemm (mode=CUBLAS_DEFAULT_MATH)
* C16: cublasGemmEx (mode=CUBLAS_DEFAULT_MATH, compute=CUBLAS_COMPUTE_16F)
* C32: cublasGemmEx (mode=CUBLAS_DEFAULT_MATH, compute=CUBLAS_COMPUTE_32F)
* R32: (mode=CUBLAS_MATH_DISALLOW_REDUCED_PRECISION_REDUCTION, compute=CUBLAS_COMPUTE_32F)
* SG: cublasSgemm (mode=CUBLAS_DEFAULT_MATH)

Latency in ms:

| GPU            | M   | N    | K     | HG        | C16       | C32       | R32       | SG        | SG/HG  | C16/HG | C32/HG | 
|----------------|-----|------|-------|-----------|-----------|-----------|-----------|-----------|--------|--------|--------| 
| V100_PCIE_16GB | 128 | 768  | 2304  | 0.033     | 0.033     | 0.054     | 0.051     | 0.067     | 2.055  | 1.006  | 1.661  | 
| V100_PCIE_16GB | 128 | 1024 | 4096  | 0.050     | 0.051     | 0.051     | 0.054     | 0.114     | 2.281  | 1.023  | 1.028  | 
| V100_PCIE_16GB | 128 | 2048 | 8192  | 0.103     | 0.104     | 0.105     | 0.119     | 0.419     | 4.086  | 1.015  | 1.024  | 
| V100_PCIE_16GB | 128 | 4096 | 16384 | 0.428     | 0.428     | 0.423     | 0.855     | 1.455     | 3.399  | 0.999  | 0.989  | 
| T4             | 128 | 768  | 2304  | 0.048     | 0.048     | 0.047     | 0.051     | 0.157     | 3.293  | 0.999  | 0.991  | 
| T4             | 128 | 1024 | 4096  | 0.077     | 0.077     | 0.0783512 | 0.0829144 | 0.345557  | 4.497  | 1.005  | 1.020  | 
| T4             | 128 | 2048 | 8192  | 0.233901  | 0.235675  | 0.235363  | 0.257197  | 1.23644   | 5.286  | 1.008  | 1.006  | 
| T4             | 128 | 4096 | 16384 | 0.994887  | 1.00059   | 1.07501   | 1.05646   | 5.59483   | 5.624  | 1.006  | 1.081  | 
| A100           | 128 | 768  | 2304  | 0.023     | 0.0234237 | 0.0252972 | 0.0252992 | 0.0501979 | 2.138  | 0.998  | 1.077  | 
| A100           | 128 | 1024 | 4096  | 0.0272268 | 0.0272073 | 0.0368818 | 0.0371932 | 0.188455  | 6.922  | 0.999  | 1.355  | 
| A100           | 128 | 2048 | 8192  | 0.0513874 | 0.0515498 | 0.0545973 | 0.109438  | 0.265237  | 5.162  | 1.003  | 1.062  | 
| A100           | 128 | 4096 | 16384 | 0.122526  | 0.119485  | 0.126554  | 0.207907  | 1.32136   | 10.784 | 0.975  | 1.033  | 
| M60            | 128 | 768  | 2304  | NA        | NA        | 0.256     | 0.253     | 0.236     | NA     | NA     | NA     | 
| M60            | 128 | 1024 | 4096  | NA        | NA        | 0.424     | 0.425     | 0.463     | NA     | NA     | NA     | 
| M60            | 128 | 2048 | 8192  | NA        | NA        | 1.496     | 1.488     | 1.710     | NA     | NA     | NA     | 
| M60            | 128 | 4096 | 16384 | NA        | NA        | 6.410     | 6.347     | 7.617     | NA     | NA     | NA     | 

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

During testing GPT-2 model in fp16, we found that the error is too large compared to PyTorch outputs. In one example, max delta in output could be 1.5 for compute type fp16, and 0.03 for compute type fp32.

### CODE REVIEW

**Comment by mrry** [comment]
@SherlockNoMad This seems related to the changes you've been prototyping on the CUDA Conv kernel in https://github.com/microsoft/onnxruntime/pull/7227.

**Comment by SherlockNoMad** [comment]
> @SherlockNoMad This seems related to the changes you've been prototyping on the CUDA Conv kernel in #7227.

This the setting mathtype in cublas, My PR was setting mathtype in cudnn.

**Review by wangyems — APPROVED** [review]
